pubsub/consumer.go

```diff
@@ -0,0 +1,198 @@
+package pubsub
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+)
+
+var (
+	// Intervals in which consumer will update heartbeat.
+	KeepAliveInterval = 30 * time.Second
+	// Duration after which consumer is considered to be dead if heartbeat
+	// is not updated.
+	KeepAliveTimeout = 5 * time.Minute
+	// Key for locking pending messages.
+	pendingMessagesKey = "lock:pending"
+)
+
+type Consumer struct {
+	id         string
+	streamName string
+	groupName  string
+	client     *redis.Client
+}
+
+type Message struct {
+	ID    string
+	Value any
```

Comment:
further thought - Producer/Consumer sending a value one way is not too useful by itself as a level. We want the consumer to process the value it got and return a result.
I think both Producer and consumer should be generics over request/response pair

Reply:
This PR is the first draft and focus here is on the protocol itself, not about the message structure. That will certainly change whether it will be concrete struct or generic, and will be in the followup PR_s. Choice of that shouldn't matter in reviewing the message queue mechanism itself.

> further thought - Producer/Consumer sending a value one way is not too useful by itself as a level. We want the consumer to process the value it got and return a result.

Same here, I'm trying to keep the PRs focused on one feature for the sake of getting changes merged soon. The result value return is sent out in a separate pr: https://github.com/OffchainLabs/nitro/pull/2197